### ISSUE TITLE
open_bitstream_unit() and uncompressed_header() use inTemporalLayer and inSpatialLayer under different conditions

### ISSUE THREAD

**Issue by wantehchang** [AOMediaCodec/av1-spec#311]
What version / commit were you testing with?
(git describe can produce this info.)

AV1 Bitstream & Decoding Process Specification
Last modified: 2019-01-08 11:48 PT
Version 1.0.0 with Errata 1

Search for "inTemporalLayer" and "inSpatialLayer" in the AV1 spec. You will find that variables by these names are used in two functions.

1) The `open_bitstream_unit()` function in Section 5.3.1 (General OBU syntax)

```
    if ( obu_type != OBU_SEQUENCE_HEADER &&
        obu_type != OBU_TEMPORAL_DELIMITER &&
        OperatingPointIdc != 0 &&
        obu_extension_flag == 1 )
    {
        inTemporalLayer = (OperatingPointIdc >> temporal_id ) & 1
        inSpatialLayer = (OperatingPointIdc >> ( spatial_id + 8 ) ) & 1
        if ( !inTemporalLayer || ! inSpatialLayer ) {
            drop_obu( )
            return
        }
    }
```

where the variable `OperatingPointIdc` is defined in the `sequence_header_obu()` function in Section 5.5.1 (General sequence header OBU syntax) as follows:

```
    operatingPoint = choose_operating_point( )
    OperatingPointIdc = operating_point_idc[ operatingPoint ]
```

2) The `uncompressed_header()` function in Section 5.9.2 (Uncompressed header syntax)

```
                    opPtIdc = operating_point_idc[ opNum ]
                    inTemporalLayer = ( opPtIdc >> temporal_id ) & 1
                    inSpatialLayer = ( opPtIdc >> ( spatial_id + 8 ) ) & 1
                    if ( opPtIdc == 0 || ( inTemporalLayer && inSpatialLayer ) ) {
                        n = buffer_removal_time_length_minus_1 + 1
                        buffer_removal_time[ opNum ]                                                 f(n)
                    }
```

Note that the conditions under which `inTemporalLayer` and `inSpatialLayer` are used are slightly different.
- In `open_bitstream_unit()`, the condition is `OperatingPointIdc != 0 && obu_extension_flag == 1`.
- In `uncompressed_header()`, the condition is `opPtIdc != 0`.

Since `OperatingPointIdc` and `opPtIdc` play the same role in the two functions, the two conditions differ in whether `obu_extension_flag == 1` is checked.

It is possible that the two conditions are equivalent because `OperatingPointIdc != 0` or `opPtIdc != 0` implies `obu_extension_flag == 1`. Even if that is true, the spec should standardize on one condition.

**Comment by peterderivaz** [comment]
There are two types of OBUs (defined in [obu-header-semantics](http://av1-draft-spec.argondesign.com/av1-spec/av1-spec.html#obu-header-semantics).  Layer-specific OBUs and not layer-specific.  The spec says "OBUs that are not layer-specific must have the obu_extension_flag set to 0."

The first piece of code is used for an arbitrary OBU and so needs to use the obu_extension_flag to decide whether the OBU is layer-specific or not.

The second piece of code is in uncompressed_header which is always a layer-specific OBU (see the table) so there is no need to check the obu_extension_flag.

In other words, I agree the second condition could indeed be changed from "( opPtIdc == 0 || ( inTemporalLayer && inSpatialLayer ) )" to "( opPtIdc == 0 || ( inTemporalLayer && inSpatialLayer && obu_extension_flag == 1) )".  However, the final condition will always be true so this could also cause trouble e.g. by confusing the reader or in making it harder to perform branch coverage testing (because the final condition is always true).

Therefore my recommendation is that no change is made.  However, this is a stylistic issue and I accept that others may disagree.

**Comment by StephenBotzko** [comment]
I agree with Peter's analysis and recommendation.

**Comment by wantehchang** [comment]
Stephen: Thank you for taking a look!

Peter: Thank you very much for the analysis. I agree with your analysis. Just wanted to confirm my understanding: The requirement you cited "OBUs that are not layer-specific must have the obu_extension_flag set to 0." does not apply to uncompressed_header because uncompressed_header is a layer-specific OBU. I believe you meant to refer to the following requirement:

> If operating_point_idc[ op ] is not equal to 0 for any value of op from 0 to operating_points_cnt_minus_1, it is a requirement of bitstream conformance that obu_extension_flag is equal to 1 for all layer-specific OBUs in the coded video sequence.

From that requirement, it follows that if opPtIdc != 0, then obu_extension_flag == 1 in uncompressed_header. Therefore it is not necessary to check obu_extension_flag == 1 in uncompressed_header. Is this correct?

I am closing this issue.

**Comment by wantehchang** [comment]
Peter: I see why you referred to the requirement "OBUs that are not layer-specific must have the obu_extension_flag set to 0." That requirement is needed to explain why open_bitstream_unit() needs to check obu_extension_flag == 1. Sorry about my confusion.

I'd still like to confirm that uncompressed_header() depends on the other requirement that I quoted. Thanks!

**Comment by peterderivaz** [comment]
Yes, I believe that is correct.

**Comment by StephenBotzko** [comment]
>>>The requirement you cited "OBUs that are not layer-specific must have the obu_extension_flag set to 0." does not apply to uncompressed_header because uncompressed_header is a layer-specific OBU. I believe you meant to refer to the following requirement:
>>>

The uncompressed header itself is not an OBU, it is a structure that is carried in OBU_FRAME_HEADER, OBU_REDUNDANT_FRAME_HEADER, and OBU_FRAME.  All three OBU types are identified as layer-specific in a table that was added to 6.4.1.  (copied below for convenience).

Stephen


obu_type
Name of obu_type
Layer-specific
0
Reserved
-
1
OBU_SEQUENCE_HEADER.
N
2
OBU_TEMPORAL_DELIMITER
N
3
OBU_FRAME_HEADER
Y
4
OBU_TILE_GROUP
Y
5
OBU_METADATA
See Table in Section 6.7.1
6
OBU_FRAME
Y
7
OBU_REDUNDANT_FRAME_HEADER
Y
8
OBU_TILE_LIST
N
9-14
Reserved
-
15
OBU_PADDING
Either

**Comment by wantehchang** [comment]
Stephen: Thank you very much for the note. I was using imprecise language for brevity, which is probably ill-advised when discussing about subtle issues in a spec.
